### ISSUE TITLE
Time dependent spy assertions

### ISSUE THREAD

**Issue by jamesrichford** [alsatian-test/alsatian#315]
Sometimes you want two functions to be called and the order of then may be important so we could add some syntax to support this. E.g.

```typescript
Expect(a.function). toHaveBeenCalled()
            .then(a.nother.function). toHaveBeenCalled();
```

@Jameskmonger @Nysosis @rlasjunies what do you guys think of this feature?

**Comment by Jameskmonger** [comment]
Not entirely keen on this if I'm honest. I think that it's probably indicative of bad design (bad test design potentially) if they need to do this.

**Comment by jamesrichford** [comment]
Generally, I would agree but there may be circumstances where this is necessary. Perhaps an example is where a function is called twice and the order of those calls are important.

e.g. 

```typescript
doThings() {
   this.logger.log("doing something");
   this._doSomething();
   this.logger.log("done something");
}
```

In this case you'd want to know that "doing something" is called before "done something" and perhaps that the actual functionality happened between the two bits of logging

**Comment by jamesrichford** [comment]
@Jameskmonger as another note I think this would be particularly important when we're talking about asynchronous functionality which is important that we can prove the order of things happens correctly when we're not writing code in a linear fashion but still expect the functionality to be executed in a given order.

**Comment by Jameskmonger** [comment]
@jamesrichford I remember at PP when I had to test things like this, it involved some mega bodges.

    var firstCall = false,
        secondCall = false;

    SpyOn(logger.log).andCall(m => {
        if (m === "doing something") {
            firstCall = true;
        } else if (m === "done something") {
            secondCall = true;
        }
    });

    Expect(firstCall).toBe(true);
    Expect(secondCall).toBe(true);

Maybe it would be good to support this. While it may not be the best thing for people to be doing, it is a genuine use case.

**Comment by jamesrichford** [comment]
@Jameskmonger exactly, this is the current sort of solution which isn't very readable but also requires you to think a lot, whereas unit tests should be super simple even in the complex cases. Perhaps if we accept this feature we should support it with documentation to say when to use it and when to refactor your code?

**Comment by Nysosis** [comment]
I've had to write tests for similar situations, and in the end created a syntax (C#) that looked similar to this (at the time it seemed to make sense):

```csharp

var reportingClient = Mock.Of<IReportingClient>();
var transaction = Mock.Of<IReportingClientTransaction>();

var sequence = new CallSequence();
reportingClient.UsingCallSequence(sequence).Setup(r => r.BeginTransaction()).Returns(transaction);
transaction.UsingCallSequence(sequence).Setup(t => t.SomeMethod());
transaction.UsingCallSequence(sequence).Setup(t => t.Commit());			transaction.UsingCallSequence(sequence).Setup(t => t.Dispose());

var sut = new Thing(reportingClient);
sut.DoMyThing();

sequence.Verify();

```

The call to `Verify()` would raise an error something along the lines of `Expected call sequence of 'BeginTransaction()', 'SomeMethod()', 'Commit()', 'Dispose()'), but instead was 'BeginTransaction()', 'Commit()', 'Dispose()'` if it wasn't in the right order. It's a little bit verbose, but got the job done.

**Comment by jamesrichford** [comment]
@Nysosis @Jameskmonger are we happy with the initial proposal here or do we want to make any changes?

**Comment by Jameskmonger** [comment]
Initial proposal LGTM @jamesrichford
